vowpalwabbit/cb_adf.cc

```diff
@@ -29,37 +29,53 @@ namespace CB_ADF
 {
 struct cb_adf
 {
-  vw* all;
-
+ public:
   cb_to_cs_adf gen_cs;
-  v_array<CB::label> cb_labels;
-  COST_SENSITIVE::label cs_labels;
-  v_array<COST_SENSITIVE::label> prepped_cs_labels;
-
-  action_scores a_s;              // temporary storage for mtr and sm
-  action_scores prob_s;           // temporary storage for sm; stores softmax values
-  v_array<uint32_t> backup_nf;    // temporary storage for sm; backup for numFeatures in examples
-  v_array<float> backup_weights;  // temporary storage for sm; backup for weights in examples
-
-  uint64_t offset;
-  bool no_predict;
+  shared_data* sd;
   bool rank_all;
```

Comment:
About rank_all, clip_p, no_predict...

If these three variables are declared in setup and then passed into the constructor, cb_adf can be constructed after args are parsed.  These three can then become private.